File: hostagent/internal/api/daily.go
package api

import (
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"orchestra/hostagent/internal/store"
	"orchestra/hostagent/internal/tasksource"
)

// handleSources lists the configured Daily task sources (Jira / Trello /
// Notion / demo).
func (s *Server) handleSources(w http.ResponseWriter, _ *http.Request) {
	all := s.currentSources().Names()
	out := make([]string, 0, len(all))
	for _, n := range all {
		if n != deliverySource { // github is a Delivery source, not a Daily one
			out = append(out, n)
		}
	}
	writeJSON(w, 200, map[string]any{"sources": out})
}

// handleSourceConfigAdd persists a task source and rebuilds the registry.
func (s *Server) handleSourceConfigAdd(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Name string `json:"name"`
		Type string `json:"type"`
	}
	if !decode(w, r, &req) {
		return
	}
	req.Type = strings.TrimSpace(req.Type)
	name := strings.TrimSpace(req.Name)
	if name == "" {
		name = req.Type
	}
	switch req.Type {
	case "jira", "trello", "notion":
	default:
		writeErr(w, 400, "type must be jira, trello or notion")
		return
	}
	if err := s.store.AddTaskSource(name, req.Type); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	s.rebuildSources()
	writeJSON(w, 201, map[string]string{"name": name, "type": req.Type})
}

// handleSourceConfigDelete removes a task source and rebuilds the registry.
func (s *Server) handleSourceConfigDelete(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	ok, err := s.store.RemoveTaskSource(name)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if !ok {
		writeErr(w, 404, "unknown source")
		return
	}
	s.rebuildSources()
	writeJSON(w, 200, map[string]string{"removed": name})
}

// handlePull performs an explicit pull from one source: fetch through the
// gateway-routed adapter, upsert into the store, and advance the incremental
// cursor. This is the "pull型" ingest — nothing is synced automatically.
func (s *Server) handlePull(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("source")
	src, ok := s.currentSources().Get(name)
	if !ok {
		writeErr(w, 404, "unknown source")
		return
	}
	cursorKey := "pull_cursor:" + name
	cursor, _ := s.store.GetState(cursorKey)
	tickets, err := src.Fetch(r.Context(), tasksource.Query{Assignee: "@me", State: "open", Since: cursor})
	if err != nil {
		writeErr(w, 502, err.Error())
		return
	}
	if err := s.store.UpsertTickets(tickets); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	for _, t := range tickets {
		if t.UpdatedAt > cursor {
			cursor = t.UpdatedAt
		}
	}
	if err := s.store.SetState(cursorKey, cursor); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	writeJSON(w, 200, map[string]any{"pulled": len(tickets), "tickets": tickets})
}

// handleRuns returns recent schedule occurrences (executed / missed) for the
// Daily run history.
func (s *Server) handleRuns(w http.ResponseWriter, r *http.Request) {
	limit := 0
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			limit = n
		}
	}
	runs, err := s.store.Runs(limit)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if runs == nil {
		runs = []store.ScheduleRun{}
	}
	writeJSON(w, 200, map[string]any{"runs": runs})
}

// handleTickets returns stored Daily tickets. With no ?source= it excludes the
// Delivery-only github source; with an explicit source it returns just that one.
func (s *Server) handleTickets(w http.ResponseWriter, r *http.Request) {
	source := r.URL.Query().Get("source")
	var ts []tasksource.Ticket
	var err error
	if source == "" {
		ts, err = s.store.TicketsExcluding(deliverySource)
	} else {
		ts, err = s.store.Tickets(source)
	}
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if ts == nil {
		ts = []tasksource.Ticket{}
	}
	writeJSON(w, 200, map[string]any{"tickets": ts})
}

type promoteReq struct {
	ID   string `json:"id"`   // pulled ticket id, e.g. "jira:PROJ-42"
	Repo string `json:"repo"` // target hostagent repo name (defaults to the ticket's)
	Base string `json:"base"` // base branch (defaults to the repo target)
}

// handlePromote turns a pulled Daily ticket into a Delivery task: it creates a
// git worktree (branch "ticket/<id>") off the base branch, so the ticket can be
// worked and reviewed through the existing Delivery flow. This is the bridge
// from Daily (external SoR, pull) to Delivery (git, worktree/PR).
func (s *Server) handlePromote(w http.ResponseWriter, r *http.Request) {
	var req promoteReq
	if !decode(w, r, &req) {
		return
	}
	ticket, err := s.store.TicketByID(req.ID)
	if err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	if ticket == nil {
		writeErr(w, 404, "unknown ticket")
		return
	}
	repoName := req.Repo
	if repoName == "" {
		repoName = ticket.Repo
	}
	repo, ok := s.repo(repoName)
	if !ok {
		writeErr(w, 400, "no target repo for ticket (pass 'repo' matching a configured repo)")
		return
	}
	base := req.Base
	if base == "" {
		base = repo.Target
	}
	branch := "ticket/" + sanitize(strings.ReplaceAll(ticket.ID, ":", "-"))
	dir := filepath.Join(os.TempDir(), "orchestra-wt", repo.Name, sanitize(branch))
	if err := s.g.AddWorktree(repo.Path, dir, branch, base); err != nil {
		writeErr(w, 500, err.Error())
		return
	}
	// Seed the task doc from the issue so the Delivery card opens with its detail
	// (title + body), editable as .orchestra/task.md and used as the agent's task.
	taskMD := "# " + ticket.Title + "\n\n" + ticket.Body + "\n"
	if ticket.URL != "" {
		taskMD += "\n<!-- orchestra: source=" + ticket.ID + " " + ticket.URL + " -->\n"
	}
	_ = s.g.WriteFile(dir, ".orchestra/task.md", taskMD)
	writeJSON(w, 201, map[string]string{
		"ticket": ticket.ID, "repo": repo.Name, "branch": branch, "worktreePath": dir,
	})
}

File: src/lib/daily.ts
// Client for the host agent's Daily task-source API: pull-model ingest of
// tickets from external systems of record (Jira / Trello / Notion). The host
// agent fetches through the security gateway, so no credentials touch the
// frontend. Delivery's git-backed tasks are separate (see hostagent.ts).

const BASE = import.meta.env.VITE_HOSTAGENT_URL ?? "http://127.0.0.1:8788";

export interface Ticket {
  id: string; // source-qualified, e.g. "jira:PROJ-42"
  source: string; // jira | trello | notion | demo
  title: string;
  body: string;
  url: string;
  state: string; // open | in_progress | closed
  repo: string;
  branch: string;
  labels: string[];
  updatedAt: string; // RFC3339
}

async function req<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(BASE + path, {
    ...init,
    headers: { "Content-Type": "application/json", ...(init?.headers ?? {}) },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error((body as { error?: string }).error ?? `HTTP ${res.status}`);
  }
  return res.json() as Promise<T>;
}

export type SourceType = "jira" | "trello" | "notion";

export interface ScheduleRun {
  id: number;
  scheduleId: string;
  name: string;
  perspective: string;
  scheduledAt: string; // RFC3339
  status: "executed" | "missed" | "failed";
  /** The directory this run produced its artifacts in. Daily runs are not git
   *  work — a schedule produces a report or a rendered video, not a branch to
   *  review — so an occurrence points at a plain directory. Empty when nothing
   *  was launched (a schedule with no template bound). */
  outputDir: string;
  containerId: string; // the sandbox whose logs are the session record (bare run)
  runId: string; // the orchestrator run id (template DAG run)
  template: string; // display label of the template used
  /** Only on occurrences recorded before Daily was separated from git. */
  repo?: string;
  branch?: string;
}

/** One file a scheduled run produced. */
export interface Artifact {
  path: string; // relative to the run's output directory
  name: string;
  size: number;
  kind: "video" | "image" | "audio" | "text" | "file";
  modTime: string;
}

export const daily = {
  /** Configured task sources (e.g. ["jira","notion"]). */
  sources: () => req<{ sources: string[] }>("/daily/sources").then((r) => r.sources),
  /** Explicitly pull from one source; returns the freshly-pulled tickets. */
  pull: (source: string) =>
    req<{ pulled: number; tickets: Ticket[] }>(
      `/daily/pull?source=${encodeURIComponent(source)}`,
      { method: "POST" },
    ),
  /** All stored pulled tickets, optionally filtered by source. */
  tickets: (source?: string) =>
    req<{ tickets: Ticket[] }>(
      `/daily/tickets${source ? `?source=${encodeURIComponent(source)}` : ""}`,
    ).then((r) => r.tickets),
  /** Recent schedule occurrences (executed / missed) for the run history. */
  runs: (limit?: number) =>
    req<{ runs: ScheduleRun[] }>(`/daily/runs${limit ? `?limit=${limit}` : ""}`).then((r) => r.runs),
  /** What a scheduled run produced, for the gallery. */
  artifacts: (runId: number) =>
    req<{ artifacts: Artifact[] }>(`/daily/artifacts?run=${runId}`).then((r) => r.artifacts),
  /**
   * URL of one artifact's bytes — used directly as a <video>/<img>/<audio> src
   * and as the download link.
   *
   * The occurrence id is the only handle: the host agent resolves it to the
   * directory that run wrote into, so no path outside it is addressable. Only
   * media is served inline; anything else comes back as a download whatever
   * this flag says.
   */
  artifactUrl: (runId: number, path: string, download = false) =>
    `${BASE}/daily/artifact?run=${runId}&path=${encodeURIComponent(path)}${download ? "&download=1" : ""}`,
  /** Remove one artifact. A gallery that only grows is one nothing can be
   *  found in, so deleting is part of reviewing rather than an admin task. */
  deleteArtifact: (runId: number, path: string) =>
    req<{ deleted: string }>(`/daily/artifact?run=${runId}&path=${encodeURIComponent(path)}`, { method: "DELETE" }),
  /** Remove an occurrence: its output directory and its history row. */
  deleteRun: (runId: number) =>
    req<{ deleted: number }>(`/daily/run?run=${runId}`, { method: "DELETE" }),
  /** Promote a pulled ticket into a Delivery worktree (branch ticket/<id>). */
  promote: (id: string, repo?: string) =>
    req<{ ticket: string; repo: string; branch: string; worktreePath: string }>(
      "/daily/promote",
      { method: "POST", body: JSON.stringify({ id, repo: repo ?? "" }) },
    ),
  /** Add a runtime-configured task source (persisted). */
  addSource: (type: SourceType, name?: string) =>
    req<{ name: string; type: string }>("/daily/sources/config", {
      method: "POST",
      body: JSON.stringify({ type, name: name ?? "" }),
    }),
  /** Remove a runtime-configured task source. */
  removeSource: (name: string) =>
    req<{ removed: string }>(`/daily/sources/config?name=${encodeURIComponent(name)}`, {
      method: "DELETE",
    }),
};
